apps/webbsite/src/libs/notion/notion.ts

```diff
@@ -0,0 +1,160 @@
+    const response = await this.notion.databases.query({
+      database_id: this.postsDatabaseID,
+      filter: {
+        property: 'Published',
+        checkbox: {
+          equals: true,
+        },
+      },
+      sorts: [
+        {
+          property: 'Slug',
+          direction: 'ascending',
+        },
+      ],
+    });
+
+    const posts = await Promise.all(
+      response.results.map(async (post: any) => {
+        if (!post) return;
+
+        const metadata = {
+          id: post.id,
+          title: post.properties.Title.title[0].plain_text,
+          author: await this.notion.users
+            .retrieve({
+              user_id: post.properties.Author.people[0].id,
+            })
+            .then((user: any) => user.name),
+          description: post.properties.Description.rich_text[0].plain_text,
+          published: post.properties.Published.checkbox,
+          slug: post.properties.Slug.formula.string,
+          tags: post.properties.Tags.multi_select.map((tag: any) => tag.name),
+          cover: post.cover.file.url,
+          dateAndTime: {
+            createdDate: formatDate(post.properties.Created.created_time),
+            createdTime: formatTime(post.properties.Created.created_time),
+            lastEditedDate: formatDate(
+              post.properties['Last edited'].last_edited_time
+            ),
+            lastEditedTime: formatTime(
+              post.properties['Last edited'].last_edited_time
+            ),
+          },
+        };
+
+        const recordMap = await this.getPostRecordMap(post.id);
+
+        return {
+          metadata,
+          recordMap,
+        };
+      })
+    );
+
+    return posts;
+  }
+
+  async getVideos() {
+    const response = await this.notion.databases.query({
+      database_id: this.videosDatabaseID,
+      filter: {
+        property: 'Published',
+        checkbox: {
+          equals: true,
+        },
+      },
+    });
+
+    const posts = await Promise.all(
+      response.results.map(async (video: any) => {
+        if (!video) return;
+
+        const metadata = {
+          id: video.id,
+          title: video.properties.Title.title[0].plain_text,
+          published: video.properties.Published.checkbox,
+          link: video.properties.Link.url,
+          tags: video.properties.Tags.multi_select.map((tag: any) => tag.name),
+          cover: video.cover.file.url,
+        };
+
+        return {
+          metadata,
+        };
+      })
+    );
+
+    return posts;
+  }
+
+  async getPostBySlug(slug: string | string[]) {
+    const posts = await this.getPosts();
+
+    const post = posts.find((post) => {
+      if (!post) return;
+
+      return post.metadata.slug === slug;
+    });
+
+    if (!post) {
+      throw new Error(`Post with slug ${slug} not found`);
+    }
+
+    return post;
+  }
+
+  async getPostRecordMap(id: string) {
+    const recordMap = await this.notionAPI.getPage(id);
+
+    return recordMap;
+  }
+}
```

Comment:
A couple points to consider for improved error handling: 

1. Error handling for the getPosts() function: The getPostBySlug function relies on the getPosts() function to retrieve the list of posts. If getPosts() throws an error, it will be unhandled and result in an unhandled rejection. To handle this, we can wrap the getPosts() call in a try-catch block.

2. Type checking for the slug parameter: The slug parameter can be either a string or an array of strings. If an array is passed, the function will not work as intended. We can add a type check to ensure that the parameter is always a string.

3. Improved error message for the posts.find() function: If the posts.find() function returns undefined, the error message will not be helpful. We can improve the error message by checking if post is defined and throwing an error if it is not.